lib/phoenix/verified_routes.ex

```diff
@@ -0,0 +1,607 @@
+
+    verify_segment(rest, route, [rewrite | acc])
+  end
+
+  defp verify_segment([_other | _], route, _acc) do
+    raise ArgumentError,
+          "verified routes require a compile-time string, got: #{Macro.to_string(route)}"
+  end
+
+  # we've reached the end of the path without finding query, return to caller
+  defp verify_segment([], _route, acc), do: {Enum.reverse(acc), _query = []}
+
+  defp verify_query(
+         [{:"::", m1, [{{:., m2, [Kernel, :to_string]}, m2, [arg]}, {:binary, _, _} = bin]}],
+         _route
+       ) do
+    {:"::", m1, [{{:., m2, [__MODULE__, :__encode_query__]}, m2, [arg]}, bin]}
+  end
+
+  defp verify_query(_other, route) do
+    raise_invalid_query(route)
+  end
+
+  defp raise_invalid_query(route) do
+    raise ArgumentError,
+          "expected query string param to be compile-time map or keyword list, got: #{Macro.to_string(route)}"
+  end
+
+  @doc """
+  Generates an integrity hash to a static asset given its file path.
+  """
+  def static_integrity(%Plug.Conn{private: %{phoenix_endpoint: endpoint}}, path) do
+    static_integrity(endpoint, path)
+  end
+
+  def static_integrity(%_{endpoint: endpoint}, path) do
+    static_integrity(endpoint, path)
+  end
+
+  def static_integrity(endpoint, path) when is_atom(endpoint) do
+    endpoint.static_integrity(path)
+  end
+
+  @doc false
+  def __encode_query__(dict) when is_list(dict) or is_map(dict) do
+    case Plug.Conn.Query.encode(dict, &to_param/1) do
+      "" -> ""
+      query_str -> "?" <> query_str
+    end
+  end
+
+  defp to_param(int) when is_integer(int), do: Integer.to_string(int)
+  defp to_param(bin) when is_binary(bin), do: bin
+  defp to_param(false), do: "false"
+  defp to_param(true), do: "true"
+  defp to_param(data), do: Phoenix.Param.to_param(data)
+
+  defp route_type(router, statics, test_path) do
+    cond do
+      static_path?(test_path, statics) -> :static
+      match_route?(router, test_path) -> :match
+      true -> :error
+    end
+  end
+
+  defp match_route?(router, test_path) when is_binary(test_path) do
+    split_path = for segment <- String.split(test_path, "/"), segment != "", do: segment
+    match_route?(router, split_path)
+  end
+
+  defp match_route?(router, split_path) when is_list(split_path) do
+    case router.__match_route__(split_path) do
+      {nil = _forward_plug, _} -> true
+      {forward_plug, _} -> match_forward_route?(router, forward_plug, split_path)
+      :error -> false
+    end
+  end
+
+  defp match_forward_route?(router, forward_router, split_path) do
+    if function_exported?(forward_router, :__routes__, 0) do
+      script_name = router.__forward__(forward_router)
+      match_route?(forward_router, split_path -- script_name)
+    else
+      true
+    end
+  end
+
+  defp build_route(route_ast, og_ast, env, endpoint_ctx, router) do
+    statics = Module.get_attribute(env.module, :statics, [])
+    {endpoint_ctx, router, statics} = Macro.expand({endpoint_ctx, router, statics}, env)
```

Comment:
This won't work because you are building a weird tuple that looks like an AST. Plus `Macro.expand` is not recursive. I also think that you already manually expand router and endpoint when they are given as arguments.

```suggestion
```